x/concentrated-liquidity/incentives_test.go

```diff
@@ -3368,3 +3378,131 @@ func (s *KeeperTestSuite) TestFindUptimeIndex() {
+				{ Weight: sdk.NewInt(1), Token:  sdk.NewCoin("bar", sdk.NewInt(150))},
+			},
+        },
+		"canonical balancer pool ID exists but pool itself is not found": {
+			// 100 existing shares and 100 shares added from balancer
+			existingClLiquidity: defaultConcentratedAssets,
+			balancerPoolAssets: defaultBalancerAssets,
+			canonicalBalancerPoolDoesNotExist: true,
+
+			expectedError: gammtypes.PoolDoesNotExistError{PoolId: invalidPoolId},
+        },
+	}
+	for name, tc := range tests {
+		s.Run(name, func() {
+			// --- Setup test env ---
+
+            s.SetupTest()
+			clPool := s.PrepareCustomConcentratedPool(s.TestAccs[0], tc.existingClLiquidity[0].Denom, tc.existingClLiquidity[1].Denom, DefaultTickSpacing, DefaultExponentAtPriceOne, sdk.ZeroDec())
+			
+			// Set up an existing full range position
+			initialLiquidity, _ := s.SetupPosition(clPool.GetId(), s.TestAccs[0], tc.existingClLiquidity[0], tc.existingClLiquidity[1], DefaultMinTick, DefaultMaxTick, s.Ctx.BlockTime())
+		
+			// If a canonical balancer pool exists, we create it and link it with the CL pool
+			balancerPoolId := s.PrepareCustomBalancerPool(tc.balancerPoolAssets, defaultBalancerPoolParams)
+			if tc.canonicalBalancerPoolDoesNotExist {
+				balancerPoolId = invalidPoolId
+			}
+
+			s.App.GAMMKeeper.SetMigrationInfo(s.Ctx, 
+				gammtypes.MigrationRecords{BalancerToConcentratedPoolLinks: []gammtypes.BalancerToConcentratedPoolLink{
+					{BalancerPoolId: balancerPoolId, ClPoolId: clPool.GetId()},
+				},
+			})
+
+			// Calculate balancer share amount for full range
+			updatedClPool, err := s.App.ConcentratedLiquidityKeeper.GetPoolById(s.Ctx, clPool.GetId())
+			qualifyingShares := math.GetLiquidityFromAmounts(updatedClPool.GetCurrentSqrtPrice(), types.MinSqrtPrice, types.MaxSqrtPrice, tc.balancerPoolAssets[1].Token.Amount, tc.balancerPoolAssets[0].Token.Amount)
+
+			if tc.canonicalBalancerPoolDoesNotExist {
+				qualifyingShares = sdk.NewDec(0)
+			}
+
+            // --- System under test ---
+
+			retrievedBalancerPoolId, addedLiquidity, err := s.App.ConcentratedLiquidityKeeper.PrepareBalancerPoolAsFullRange(s.Ctx, clPool.GetId())
+
+			// --- Assertions ---
+
+			if tc.expectedError != nil {
+				s.Require().Error(err)
+                s.Require().ErrorContains(err, tc.expectedError.Error())
+
+				// Ensure that returned balancer pool ID is correct
+				s.Require().Equal(uint64(0), retrievedBalancerPoolId)
+            } else {
+				s.Require().NoError(err)
+
+				// Ensure that returned balancer pool ID is correct
+				s.Require().Equal(balancerPoolId, retrievedBalancerPoolId)
+			}
+
+			updatedClPool, err = s.App.ConcentratedLiquidityKeeper.GetPoolById(s.Ctx, clPool.GetId())
+			s.Require().NoError(err)
+
+			clPoolUptimeAccumulators, err := s.App.ConcentratedLiquidityKeeper.GetUptimeAccumulators(s.Ctx, clPool.GetId())
+			s.Require().NoError(err)
+
+			for _, uptimeAccum := range clPoolUptimeAccumulators {
```

Comment:
Hmm this is a strong guarantee at a different level of abstraction (fixed number of accums are set up at pool creation). Will add the check anyway but in general this would be extraneous